File: ServerCore/src/main/java/org/elkoserver/foundation/server/HttpConnectionSetup.java
package org.elkoserver.foundation.server;

import org.elkoserver.foundation.net.JSONHTTPFramer;
import org.elkoserver.foundation.net.MessageHandlerFactory;
import org.elkoserver.foundation.net.NetAddr;
import org.elkoserver.foundation.net.NetworkManager;
import org.elkoserver.foundation.properties.ElkoProperties;
import org.elkoserver.foundation.server.metadata.AuthDesc;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

import java.io.IOException;

class HttpConnectionSetup extends BaseConnectionSetup {
    private final String domain;
    private final NetworkManager myNetworkManager;
    private final MessageHandlerFactory actorFactory;
    private final String rootURI;
    private final String serverAddress;

    HttpConnectionSetup(String label, String host, AuthDesc auth, boolean secure, ElkoProperties props, String propRoot, NetworkManager networkManager, MessageHandlerFactory actorFactory, Trace trServer, Trace tr, TraceFactory traceFactory) {
        super(label, host, auth, secure, props, propRoot, trServer, tr, traceFactory);

        domain = determineDomain(host, props, propRoot);
        myNetworkManager = networkManager;
        this.actorFactory = actorFactory;
        rootURI = props.getProperty(propRoot + ".root", "");
        serverAddress = host + "/" + rootURI;
    }

    private String determineDomain(String host, ElkoProperties props, String propRoot) {
        String result = props.getProperty(propRoot + ".domain");

        if (result == null) {
            int colonIndex = host.indexOf(':');
            if (colonIndex == -1) {
                result = host;
            } else {
                result = host.substring(0, colonIndex);
            }
            int indexOfLastDot = result.lastIndexOf('.');
            indexOfLastDot = result.lastIndexOf('.', indexOfLastDot - 1);
            if (indexOfLastDot > 0) {
                result = result.substring(indexOfLastDot + 1);
            }
        }

        return result;
    }

    @Override
    public String getProtocol() {
        return "http";
    }

    @Override
    public String getServerAddress() {
        return serverAddress;
    }

    @Override
    NetAddr tryToStartListener() throws IOException {
        return myNetworkManager.listenHTTP(
                bind,
                actorFactory,
                msgTrace, secure, rootURI,
                new JSONHTTPFramer(msgTrace, traceFactory));
    }

    @Override
    String getListenAddressDescription() {
        return host + "/" + rootURI + "/ in domain " + domain;
    }

    @Override
    String getValueToCompareWithBind() {
        return host;
    }
}

File: ServerCore/src/main/kotlin/org/elkoserver/foundation/server/HttpConnectionSetup.kt
package org.elkoserver.foundation.server

import org.elkoserver.foundation.net.JSONHTTPFramer
import org.elkoserver.foundation.net.MessageHandlerFactory
import org.elkoserver.foundation.net.NetworkManager
import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.foundation.server.metadata.AuthDesc
import org.elkoserver.util.trace.Trace
import org.elkoserver.util.trace.TraceFactory
import java.io.IOException

internal class HttpConnectionSetup(label: String?, host: String, auth: AuthDesc?, secure: Boolean, props: ElkoProperties, propRoot: String, private val myNetworkManager: NetworkManager, private val actorFactory: MessageHandlerFactory, trServer: Trace?, tr: Trace?, traceFactory: TraceFactory?) : BaseConnectionSetup(label, host, auth!!, secure, props, propRoot, trServer!!, tr!!, traceFactory!!) {
    private val domain: String?
    private val rootURI: String
    override val serverAddress: String

    override val protocol = "http"

    @Throws(IOException::class)
    override fun tryToStartListener() =
            myNetworkManager.listenHTTP(bind, actorFactory, msgTrace, secure, rootURI, JSONHTTPFramer(msgTrace, traceFactory))

    override val listenAddressDescription: String
        get() = "$host/$rootURI/ in domain $domain"

    override val valueToCompareWithBind: String
        get() = host

    init {
        domain = determineDomain(host, props, propRoot)
        rootURI = props.getProperty("$propRoot.root", "")
        serverAddress = "$host/$rootURI"
    }

    companion object {
        private fun determineDomain(host: String, props: ElkoProperties, propRoot: String): String? {
            var result = props.getProperty("$propRoot.domain")
            if (result == null) {
                val colonIndex = host.indexOf(':')
                result = if (colonIndex == -1) {
                    host
                } else {
                    host.substring(0, colonIndex)
                }
                var indexOfLastDot = result.lastIndexOf('.')
                indexOfLastDot = result.lastIndexOf('.', indexOfLastDot - 1)
                if (indexOfLastDot > 0) {
                    result = result.substring(indexOfLastDot + 1)
                }
            }
            return result
        }
    }
}